### PR TITLE
Prevent viewport.showOnScreen from scrolling the viewport if the specified Rect is already visible. 

### LINKED ISSUE #25507: CustomScrollView and TextField issue
## Steps to Reproduce
So having a `TextField` within a `CustomScrollView` causes the entire view to scroll to the top once the keyboard comes up.

``` dart
Scaffold(
	body: CustomScrollView(
		slivers: <Widget>[
			SliverList(
				delegate: SliverChildListDelegate(
					<Widget>[
						Align(
							alignment: Alignment.topLeft,
							child: FlutterLogo(size: 64.0),
						),
						ListTile(
							title: Text('Some Text'),
						),
						ListTile(),
					],
				),
			),
			SliverAppBar(
				backgroundColor: Theme.of(context).scaffoldBackgroundColor,
				elevation: 0.0,
				automaticallyImplyLeading: false,
				pinned: true,
				floating: false,
				title: TextField(
					style: Theme.of(context).primaryTextTheme.title,
					decoration: InputDecoration(
						border: OutlineInputBorder(
								borderRadius: BorderRadius.all(Radius.circular(8.0))),
						suffixIcon: Icon(Icons.search),
						hintText: '검색',
					),
				),
			),
			SliverList(
				delegate: SliverChildListDelegate(List.generate(
						100,
						(i) => ListTile(
									title: Text('Scroll'),
								)).toList()),
			),
		],
	),
);
```

## Logs

```
[✓] Flutter (Channel master, v1.1.3-pre.11, on Mac OS X 10.14.2 18C54, locale en-NG)
    • Flutter version 1.1.3-pre.11 at /Users/apple/Desktop/flutter
    • Framework revision c5251cdc02 (4 days ago), 2018-12-14 16:09:17 -0800
    • Engine revision 101b27da7a
    • Dart version 2.1.0 (build 2.1.0-dev.9.4 f9ebf21297)

[✓] Android toolchain - develop for Android devices (Android SDK 28.0.3)
    • Android SDK at /Users/apple/Library/Android/sdk
    • Android NDK at /Users/apple/Library/Android/sdk/ndk-bundle
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = /Users/apple/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 10.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.1, Build version 10B61
    • ios-deploy 1.9.4
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 24.2.1
    • Dart plugin version 173.4700
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] VS Code (version 1.30.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 2.21.1

[✓] Connected device (1 available)
    • SM G955F • ce041714616c00130c • android-arm64 • Android 7.0 (API 24)

• No issues found!
```

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

- ~Check if the viewport needs to scroll at all in `RenderViewport.showOnScreen` before calculating the target scroll offset. The check is done via paint transform so it does not make assumptions about the slivers.~
- Check if the outmost sliver has a greater-than-zero `maxScrollObstructionExtent`.
- Trim the `Rect` in `RenderSliverPersistentHeader` (and its subclasses) to prevent `showOnScreen` from unpinning the header in the viewport.
- Expands the `RenderSliverFloatingPersistentHeader` and `RenderSliverFloatingPinnedPersistentHeader` when instructed by `showOnScreen`.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/25507

## Tests

I added the following tests:

A pinned persistent header should not scroll when its descendant EditableText gains focus, etc.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [x] I wrote a design doc: https://docs.google.com/document/d/1BZhxy176uUnqOCnXdnHM1XetS9mw9WIyUAOE-dgVdUM/edit?usp=sharing
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -1686,13 +1686,11 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
       }
-      final Rect inflatedRect = Rect.fromLTRB(
-          newCaretRect.left - widget.scrollPadding.left,
-          newCaretRect.top - widget.scrollPadding.top,
-          newCaretRect.right + widget.scrollPadding.right,
-          newCaretRect.bottom + bottomSpacing,
-      );
-      _editableKey.currentContext.findRenderObject().showOnScreen(
+
+      final Rect inflatedRect = widget.scrollPadding
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:1690]: An alternative to the trimming in `RenderSliverPersistentHeader.showOnScreen` is to confine `inflatedRect` so that it doesn't exceed `renderEditable`'s paintBounds. But that breaks around 6 tests.

File: packages/flutter/lib/src/rendering/viewport.dart [hunk 2]
```diff
@@ -783,7 +783,6 @@ abstract class RenderViewportBase<ParentDataClass extends ContainerParentDataMix
     final double offsetDifference = offset.pixels - targetOffset;
 
     final Matrix4 transform = target.getTransformTo(this);
-    applyPaintTransform(child, transform);
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/viewport.dart:786]: It seems the paint transform that does `child -> this` is applied twice here.

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 3]
```diff
@@ -196,6 +196,68 @@ abstract class RenderSliverPersistentHeader extends RenderSliver with RenderObje
...
+      ? MatrixUtils.transformRect(descendant.getTransformTo(child), rect ?? descendant.paintBounds)
+      : null;
+
+    // Trims the part of `rect` that protrudes `child`'s leading edge.
+    //
+    // If the `rect` a descendant specified in its showOnScreen call exceeds
+    // the leading edge of this sliver (which is usually the same as that of
+    // `child`), the viewport will move towards the leading edge (reduce its
+    // scroll offset) to unpin the persistent header. This is almost always
+    // undesirable.
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart:217]: It took me awhile to understand this comment, so I tried to clarify a bit. Definitely double check this and make sure I understood correctly, though.

> Trims the given Rect `original` so that it fits within the boundaries given by `top`, `right`, `bottom`, and `left`.
> 
> This is used to prevent the case where `rect` or `descendant` specified in showOnScreen exceed the leading edge of this sliver.  If this were to happen, the viewport would move towards the leading edge (reducing its scroll offset) in order to unpin the persistent header.  This is almost always undesirable.

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 4]
```diff
@@ -585,6 +647,61 @@ abstract class RenderSliverFloatingPersistentHeader extends RenderSliverPersiste
...
+    }
+
+    minTargetExtent = minTargetExtent.clamp(childExtent, maxExtent) as double;
+    // Expands the header if needed, with animation if possible.
+    if (minTargetExtent > childExtent) {
+      if (snapConfiguration != null) {
+        _controller ??= AnimationController(vsync: snapConfiguration.vsync, duration: duration);
+        _controller.duration = duration;
+        _animation = _controller
+        .drive(
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart:685]: This should be indented according to the style guide, if I'm not mistaken.

**Review by justinmc — COMMENTED** [review]
LGTM from a quick review, but I'm no sliver expert, so I'll definitely defer to other reviewers.

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart (hunk 3 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/sliver_persistent_header.dart:217]: :+1:  I'll put the comment above the `showOnScreen` method instead. 

Is it OK to limit the size of the caret's scroll padding instead, so that it will never exceed the leading edge (see the comment in "editable_text.dart" file)? That makes a little bit more sense to me (in case the caller of `showOnScreen` really expects us to respect the `rect` specified). Or maybe we should reduce the default `scrollPadding` (`EdgeInsets.all(20.0)`) and document this edge case?

Comment (justinmc) [reply packages/flutter/lib/src/rendering/sliver_persistent_header.dart:217]: That makes sense, but would it be a breaking change in normal situations?

File: packages/flutter/lib/src/rendering/viewport.dart (hunk 2 above)
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/viewport.dart:786]: Nice catch. Did you add a test for this?

File: packages/flutter/test/rendering/viewport_test.dart [hunk 5]
```diff
@@ -986,6 +1006,211 @@ void main() {
...
+        ),
+      );
+
+      controller.jumpTo(300.0 * 15);
+      await tester.pumpAndSettle();
+
+      final Finder pinnedHeaderContent = find.descendant(
+        of: find.byWidget(children[10], skipOffstage: false),
+        matching: find.byKey(headerKey, skipOffstage: false),
+        skipOffstage: false,
```
Comment (goderbauer) [inline packages/flutter/test/rendering/viewport_test.dart:1050]: If the header is still visible (as said below), why do you need skipOffstage: false?

File: packages/flutter/test/rendering/viewport_test.dart [hunk 6]
```diff
@@ -986,6 +1006,211 @@ void main() {
...
+              ),
+            ),
+          ),
+        ),
+      );
+
+      final Finder pinnedHeaderContent = find.descendant(
+        of: find.byWidget(children[10], skipOffstage: false),
+        matching: find.byKey(headerKey, skipOffstage: false),
+        skipOffstage: false,
```
Comment (goderbauer) [inline packages/flutter/test/rendering/viewport_test.dart:1112]: same here.

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 7]
```diff
@@ -585,6 +647,61 @@ abstract class RenderSliverFloatingPersistentHeader extends RenderSliverPersiste
...
+        minTargetExtent = childBounds?.bottom ?? childExtent;
+        break;
+      case AxisDirection.left:
+        minTargetExtent = childExtent - (childBounds?.left ?? 0);
+        break;
+    }
+
+    minTargetExtent = minTargetExtent.clamp(childExtent, maxExtent) as double;
+    // Expands the header if needed, with animation if possible.
+    if (minTargetExtent > childExtent) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart:680]: Maybe I missed it, but did you have tests for when the header animates and for when it doesn't?

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 8]
```diff
@@ -196,6 +196,68 @@ abstract class RenderSliverPersistentHeader extends RenderSliver with RenderObje
...
+    Rect rect,
+    Duration duration = Duration.zero,
+    Curve curve = Curves.ease,
+  }) {
+    assert(child != null);
+    final Rect childBounds = descendant != null
+      ? MatrixUtils.transformRect(descendant.getTransformTo(child), rect ?? descendant.paintBounds)
+      : null;
+
+    // Trims the part of `rect` that protrudes `child`'s leading edge.
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/sliver_persistent_header.dart:211]: You're right we only need to override `showOnScreen` for pinned headers.  But I'm not sure which approach we should take, relying on the caller to pass us the right rect, or trim the rect the caller passed down? The latter sounds a bit hacky to me.

File: packages/flutter/lib/src/rendering/viewport.dart [hunk 9]
```diff
@@ -986,16 +985,61 @@ abstract class RenderViewportBase<ParentDataClass extends ContainerParentDataMix
...
+        canSkipScrolling = canSkipScrolling && localRect.bottom <= size.height;
+        break;
+      case AxisDirection.left:
+        canSkipScrolling = canSkipScrolling && localRect.left >= 0;
+        canSkipScrolling = canSkipScrolling && localRect.right <= size.width - extentOfPinnedSlivers;
+        break;
+    }
+
+    if (!canSkipScrolling) {
+      localRect = RenderViewportBase.showInViewport(
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/viewport.dart:1031]: `showInViewport` uses `getOffsetToReveal` (which assumes the target sliver moves linearly) to determine whether a sliver is still visible. With pinned headers (floating headers) we can get false negatives (the scroll offset indicates the sliver is not in the viewport but it is, or it takes less scroll offset than `getOffsetToReveal` suggested to reveal).

File: packages/flutter/lib/src/rendering/viewport.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/viewport.dart]: Removed. Adding this in a different PR.

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 10]
```diff
@@ -17,6 +21,25 @@ import 'sliver.dart';
 import 'viewport.dart';
 import 'viewport_offset.dart';
 
+// Trims the specified edges of the given `Rect` [original], so that they do not
+// exceed the given values.
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart:25]: I'm tempted to make this a private extension method to avoid the null check. Is using method extensions not recommended?

**Comment by LongCatIsLooong** [comment]
@goderbauer @justinmc @Piinks updated. Could you take a look?

**Review by justinmc — APPROVED** [review]
LGTM, just a bunch of nits.  This approach looks pretty clean to me, but I'll continue to defer to @goderbauer.

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 11]
```diff
@@ -39,6 +62,86 @@ class OverScrollHeaderStretchConfiguration {
...
+  /// [RenderBox], or that of the persistent header when `child` is null.
+  /// {@endtemplate}
+  ///
+  /// Defaults to true and must not be null.
+  final bool ignoreLeading;
+
+  /// {@template flutter.rendering.persistentHeader.minShowOnScreenExtent}
+  /// The smallest the floating header can expand to in the main axis direction,
+  /// when `showOnScreen` is called, in addition to the persistent header's
+  /// `minExtent`.
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart:108]: How is this different from `PersistentHeaderDelegate.minExtent`?

Comment (goderbauer) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart:108]: What `minExtent are you refering to here? Is this https://master-api.flutter.dev/flutter/widgets/SliverPersistentHeaderDelegate/minExtent.html? Maybe make the doc link to it properly. Same for maxExtent

File: packages/flutter/lib/src/material/app_bar.dart [hunk 12]
```diff
@@ -981,6 +992,9 @@ class SliverAppBar extends StatefulWidget {
     this.stretchTriggerOffset = 100.0,
     this.onStretchTrigger,
     this.shape,
+    this.preventShowOnScreenUnpinning = true,
+    this.minShowOnScreenExtent,
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/app_bar.dart:996]: These two (maybe all three) new properties seem to be on a different abstraction level then the other properties.

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 13]
```diff
@@ -585,6 +763,80 @@ abstract class RenderSliverFloatingPersistentHeader extends RenderSliverPersiste
...
+
+    // A stretch header can have a bigger childExtent than maxExtent.
+    final double effectiveMaxExtent = math.max(childExtent, maxExtent);
+    double minTargetExtent;
+    Rect targetRect;
+    switch (applyGrowthDirectionToAxisDirection(constraints.axisDirection, constraints.growthDirection)) {
+      case AxisDirection.up:
+        minTargetExtent = childExtent - (childBounds?.top ?? 0);
+        targetRect = showOnScreenConfiguration.ignoreLeading
+          ? _trim(childBounds, bottom: childExtent) : childBounds;
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart:795]: nit: format so that the `: childBounds` is on a separate line aligned with the `?`

same below

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 14]
```diff
@@ -585,6 +763,80 @@ abstract class RenderSliverFloatingPersistentHeader extends RenderSliverPersiste
...
+          ? _trim(childBounds, right: childExtent) : childBounds;
+        break;
+    }
+
+    minTargetExtent = minTargetExtent
+      .clamp(
+        showOnScreenConfiguration.minShowOnScreenExtent ?? childExtent,
+        math.max(showOnScreenConfiguration.maxShowOnScreenExtent ?? effectiveMaxExtent, childExtent),
+      )
+      .clamp(childExtent, effectiveMaxExtent) as double;
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart:819]: This double-clamp is hard to follow. Maybe calculate the lowerLimit/upperLimit and then clamp once?

File: packages/flutter/lib/src/material/app_bar.dart [hunk 15]
```diff
@@ -981,6 +992,9 @@ class SliverAppBar extends StatefulWidget {
     this.stretchTriggerOffset = 100.0,
     this.onStretchTrigger,
     this.shape,
+    this.preventShowOnScreenUnpinning = true,
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/material/app_bar.dart:995]: The use case doesn't seem common at all. I'll remove it.

**Comment by LongCatIsLooong** [comment]
@justinmc @goderbauer updated the API and the documentation. Should be ready for another round of review.

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 16]
```diff
@@ -39,6 +58,61 @@ class OverScrollHeaderStretchConfiguration {
...
+  /// larger than that maximum extent, it will remain unchanged.
+  ///
+  /// This parameter can be set to the persistent header's `minExtent` (or
+  /// `double.negativeInfinity`) so the persistent header will never try to
+  /// expand when [RenderObject.showOnScreen] is called on it.
+  ///
+  /// Defaults to null, in which case no additional constraints are applied.
+  /// Must be greater than or equal to [minShowOnScreenExtent] if it is also not
+  /// null. Has no effect unless the persistent header is a floating header.
+  /// {@endtemplate}
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart:112]: Lone `endtemplate` here with no `template`, unless I'm missing it...

**Review by justinmc — APPROVED** [review]
Quick doc problem I found, but otherwise LGTM 👍

File: packages/flutter/lib/src/rendering/viewport.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/viewport.dart:1058]: The added logic does assume the layout is up to date with the current `offset`. I'll add a check to invalidate the skip logic if the assumption does not hold.

[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/viewport.dart]: @goderbauer got rid of the visibility check because the viewport could be scrolling / about to scroll. Could you take a look?

[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/viewport.dart]: Sorry nvm that broke some tests. Fixing.

**Comment by LongCatIsLooong** [comment]
I updated the logic in `getOffsetToReveal` & `revealInViewport`, so that it works even if the developer calls `offset.jumpTo` right before `RenderViewport.showOnScreen`, which I feel is not a rare use case. An alternative approach that supports this use case is to suggest putting `showOnScreen` in a post frame callback in the API docs, but since we're already assuming slivers move linearly in the viewport(except for persistent headers), this feels like an overkill.

I did have to special-case persistent headers in `getOffsetToReveal` however (i.e. making the assumption that if a sliver's maxScrollObstructionExtent > 0 then it will be pinned to the leading edge when `viewport.offset` increases).

@justinmc @goderbauer @Piinks could you please take a look at this again?

**Review by justinmc — APPROVED** [review]
LGTM but still deferring to others for sliver knowledge.

Your approach for getOffsetToReveal and revealInViewport seems like the right approach the way you describe it.

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 17]
```diff
@@ -394,15 +509,22 @@ class FloatingHeaderSnapConfiguration {
   /// Creates an object that specifies how a floating header is to be "snapped"
   /// (animated) into or out of view.
   FloatingHeaderSnapConfiguration({
-    @required this.vsync,
+    @Deprecated(
+      'Specify SliverPersistentHeaderDelegate.vsync instead. '
+      'This feature was deprecated after v1.19.0.'
+    )
+    this.vsync,  // ignore: deprecated_member_use_from_same_package
```
Comment (Piinks) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart:516]: I think we can remove this ignore, the analysis settings don't require it anymore.
```suggestion
    this.vsync,
```

**Review by Piinks — APPROVED** [review]
This LGTM. There may be some implications to consider later on in NestedScrollView, which currently does not support floating && snapping, just floating xor snapping. That's being tracked in https://github.com/flutter/flutter/issues/59189
